[CPU][SVE] Update default tiles sizes for matmul ops

banach-space opened the pull request:
The default tile sizes for SVE for matmuls, [8, 32, 16], are basically a copy of one of the existing configurations. In the case of SVE, that configuration leads to too aggressive unrolling and poor performance due to register spilling. Hence the need to update.

This patch updates the default tile sizes for SVE to [8, 16, 1]. The middle dimension corresponds to vector sizes after vectorisation (that's also the dimension that's configured to be scalable). As the base vector register size for SVE is 128 bits, there will be (depending on the element size):

  * (16 x vscale) elements per vector register for i8,
  * (16 / 2 x vscale) elements per vector registers for i16,
  * (16 / 4 x vscale) elements per vector registers for i32,
  * (...)

So, effectively, 16 is the lowest number that can be used to avoid under utilisation of vector registers (i.e. a lower number might be fine for wider elements, but not for i8).

As for the remaining tile sizes, those were determined experimentally by benchmarking `linalg.matmul` for:

  * square matrices (1020x1020, 1021x1021, 1024x1024),
  * both i8 and f32 element types,
  * input tensors with static and dynamic shapes.

In all the of the above cases, the new configuration improves the performance.

It's probably worth pointing out that during compilation, IREE will reduce the leading tile size from 8 to 6. So, effectively, the tile sizes will be [6, 16, 1].

banach-space committed:
[CPU][SVE] Update default tiles sizes for matmul ops

The default tile sizes for SVE for matmuls, [8, 32, 16], are basically a
copy of one of the existing configurations. In the case of SVE, that
configuration leads to too aggressive unrolling and poor performance due
to register spilling. Hence the need to update.

This patch updates the default tile sizes for SVE to [8, 16, 1]. The
middle dimension corresponds to vector sizes after vectorisation (that's
also the dimension that's configured to be scalable). As the base vector
register size for SVE is 128 bits, there will be (depending on the
element size):

  * (16 x vscale) elements per vector register for i8,
  * (16 / 2 x vscale) elements per vector registers for i16,
  * (16 / 4 x vscale) elements per vector registers for i32,
  * (...)

So, effectively, 16 is the lowest number that can be used to avoid under
utilisation of vector registers (i.e. a lower number might be fine for
wider elements, but not for i8).

As for the remaining tile sizes, those were determined experimentally by
benchmarking `linalg.matmul` for:

  * square matrices (1020x1020, 1021x1021, 1024x1024),
  * both i8 and f32 element types,
  * input tensors with static and dynamic shapes.

In all the of the above cases, the new configuration improves the
performance.

It's probably worth pointing out that during compilation, IREE will
reduce the leading tile size from 8 to 6. So, effectively, the tile
sizes will be [6, 16, 1].

MacDue reviewed (commented):
You'll need to update the tests in `materialize_aarch64_launch_configuration.mlir`:

To `[16]` here:
https://github.com/openxla/iree/blob/main/compiler/src/iree/compiler/Codegen/LLVMCPU/test/materialize_aarch64_launch_configuration.mlir#L101

It may also be good to update the sizes the test below, as it'd be less clear this is testing the snapping to powers-of-2 if `[16]` is the default size:
https://github.com/openxla/iree/blob/main/compiler/src/iree/compiler/Codegen/LLVMCPU/test/materialize_aarch64_launch_configuration.mlir#L142

banach-space committed:
fixup! [CPU][SVE] Update default tiles sizes for matmul ops

Update tests

banach-space committed:
fixup! [CPU][SVE] Update default tiles sizes for matmul ops

Update test so that the tile size is reduced to match the nearest power
of 2 - that's what the updated test was intended to verify in the first
place.

MaheshRavishankar reviewed (commented):
I dont expect this to cause any regressions, but should run benchmarks for things that touch these. So waiting on confirmation from benchmark runs.

MaheshRavishankar reviewed (approved):
Looks good to me. Thanks!

MacDue reviewed (approved):
LGTM 👍

hanhanW reviewed (approved):
6 is intended for legacy issues. We should be able to move out of the state today.

It is set to `6` because we were not able to vectorize them. Today we have much more tricks, and we should be able to support them.
